llvm-passes/ObfuscateVar/ObfuscateVar.cpp

```diff
@@ -0,0 +1,365 @@
+//Compile : cmake -DLLVM_ROOT=$HOME/Documents/Programmation/Obfuscation/llvm/build ..
+#include <map>
+#include <utility> //std::pair, std::make_pair
+
+
+#include "llvm/Pass.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/IR/ValueMap.h"
+
+#include "llvm/Support/raw_ostream.h"
+
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/IPO/PassManagerBuilder.h"
+
+#ifndef NDEBUG
+#include "llvm/IR/Verifier.h"
+#include "llvm/Support/Debug.h"
+#endif
+
+using namespace llvm;
+
+namespace {
+  typedef uint32_t typeInter;
+  typedef std::pair<Value*,Value*> typeMapValue;
+  typedef ValueMap<Value*,typeMapValue> typeMap;
+  
+  class ObfuscateVar : public BasicBlockPass {
+  public:
+  
+    static char ID;
+    ObfuscateVar() : BasicBlockPass(ID) {}
+
+    virtual bool runOnBasicBlock(BasicBlock &BB) override {
+      /* 
+         From https://www.cs.ox.ac.uk/files/2936/RR-10-02.pdf:
+         Given a variable X, we split it such as
+         - A = X%10
+         - B = X//10
+       
+         Z=X+Y will be transform into :
+         - Z_A = (X_A+Y_A) mod 10
+         - Z_B = { 10*(X_B+Y_B)+(X_A+Y_A) } div 10
+
+         To rebuild the variable : 
+         X = 10*X_B+X_A
+
+      */
+    
+      for (typename BasicBlock::iterator I = BB.getFirstInsertionPt(),end = BB.end(); I != end; ++I) {
+      
+        Instruction &Inst = *I;
+        bool isVolatile = false;
+      
+        if(isValidInstForSplit(Inst)) {//Check if the instruction can be splited
+        
+          for (size_t i=0; i < Inst.getNumOperands(); ++i) {
+            if (Value *V = isValidCandidateOperand(Inst.getOperand(i))) {
+
+              if(!isSplited(V)){//Check if the operand is splited
+                dbgs() << *V << " isn't splitted\n";
+                splitVariable(V,Inst);
+                //dbgs() << *V << " is now splited\n";
+                
+              }
+
+            }//isValidCandidateOperand
+          }//for
+
+          if(BinaryOperator *Binop = dyn_cast<BinaryOperator>(&Inst)) {
+
+            //Get the operator's operands
+            Value *op0 = parseOperand(Binop->getOperand(0));
+            Value *op1 = parseOperand(Binop->getOperand(1));
+           
+            if(!(isSplited(op0) && isSplited(op1))){//Check if the two operands are splited. It should be ok due to the previous loop
+              dbgs() << "Error : operands aren't splited !\n";
+              break;
+            }
+           
+            switch(Binop->getOpcode()){
+              
+            case Instruction::Add :{//Split add instruction
+
```

Comment:
maybe but this block into a function? That would make it easier to add more operation support